**1 Sample Z test or 1 Sample t test:**
**Example:**

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

In [ ]:
import statsmodels

In [ ]:
from statsmodels import stats

In [ ]:
from statsmodels.stats import weightstats

In [ ]:
x = [3,7,6,5,9,8,2,3,5,4,6,7]

In [ ]:
np.mean(x)

5.416666666666667

Iam claiming that the mean of x is morethan 6.5

H0 a >= 6.5

HA a < 6.5

small eg: cylinder lasts for atleast 40 days.

H0 a >= 40
HA a < 40

In [ ]:
statsmodels.stats.weightstats.ztest(x,value=6.5,alternative='smaller')

#output is (z or test statistic,  p- value)

(-1.744291601622091, 0.04055412767721645)

here p < LOS
0.0405 < 0.05


Hence we reject the null hypothesis.
We need to make some changes.
The difference is significant.

This is not possible in realtime. We don't change.
This is just for understanding.

let's say we change the null hypothesis is changed to

Iam claiming that the mean of x is morethan 6.5

H0 a >= 6.5

HA a < 6.5

In [ ]:
statsmodels.stats.weightstats.ztest(x,value=6,alternative='smaller')

(-0.9392339393349719, 0.17380532364640544)

here p > LOS

0.17 > 0.05

but mean = 5.41 < 6.0

but p > LOS, can be attributed to the chance variation.


**V.V.IMP**

--> Chance variation and Bias cannot be quantified. We can't measure Chance variation and Bias.

--> p-value is the prob of committing type I error.

**V.V.IMP**


--> if p < LOS (0.05), that means the diff between SS and PP is large enough not to be ignored, So action is required.

--> The diff is not only due to chance variation,bias. there is an actual difference.



In [ ]:
# 1 Step deeper for understanding

np.std(x,ddof=1)   #finding std deviation


2.151461800448216

In [ ]:
len(x)

12

In [ ]:
#z value

(5.416-6.0)/(2.151/12**0.5)

-0.9405092251234065

In [ ]:
(5.416-6.5)/(2.151/12**0.5)

-1.7457397260852276

In [ ]:
from scipy.stats import norm

In [ ]:
norm.ppf(-1.774)

nan

In [ ]:
norm.cdf(-1.744)

0.04057954491888611

In [ ]:
norm.cdf(-0.9405092251234065)

#p value for claimed value 6.5

0.17347820997037183

Going deeper for chance variation:


5.41 is the sample mean and because error is on left side, we would prefer this to be as high as possible.

The highest possible value at 95% Confidence level, is called the 95% upper bound.

In [ ]:
norm.ppf(0.05)

#error is only on one side. i.e., left side

-1.6448536269514729

In [ ]:
5.41 + 1.645*(2.151461800448216/12**0.5)   #sample mean + z * (sigma/sqrt(n))

#the mean can go until 6.43 but not 6.44


6.431665948328879

We cannot say statistical siginificance without p-value.

The mean can go until 6.43 but not > 6.43

--> If we increase mean > 6.43, then p - value will be < 0.05

--> If we decrease mean < 6.43, then p - value will be > 0.05

95% upper bound formula:

upper bound = sample mean + Z * sigma/sqrt(n)

lower bound = sample mean - Z * sigma/sqrt(n)



**2 SAMPLE t test**

--> 2 SAMPLE t test  is used to compare the differences of the averages of 2 independent samples with the claim value and establish whether the difference is statistically significant or not.

--> predictor is categorical, response is continuous.

In [ ]:
x = [3,7,6,5,9,8,2,3,5,4,6,7]

In [ ]:
y = [9,6,8,10,6,4,7,8,9,10]

**claim:**

H0 mean(y) - mean(x) >=2.7

Ha mean(y) - mean(x)  < 2.7 (left tail test)



In [ ]:
np.mean(y) - np.mean(x)

2.283333333333333

In [ ]:
statsmodels.stats.weightstats.ztest(y,x,value=2.7,alternative='smaller')

(-0.47198804896079605, 0.3184676595008745)

here p > LOS, the null hypothesis is not rejected.
The statistical difference is not significant.

the differnce between 2.28 and 2.7 can be ignored.

In [ ]:
statsmodels.stats.weightstats.ztest(y,x,value=4.2,alternative='smaller')

(-2.1711450252196607, 0.014960105716979144)

here we changed null hypothesis
H0 mean(y) - mean(x) >=4.2

Ha mean(y) - mean(x) < 4.2


here in the solution p< LOS.
THE diff between 2.28 and 4.2 cannot be ingored.
The difference is significant.
change is requied.

In [ ]:
x = [3,7,6,5,9,8,2,3,5,4,6,7]

In [ ]:
y = [9,6,8,10,6,4,7,8,9,10]

RIGHT TAIL TEST:

claim:

H0 mean(y) - mean(x) <=1.5

Ha mean(y) - mean(x) > 1.5 (right tail test)


In [ ]:
statsmodels.stats.weightstats.ztest(y,x,value=1.5,alternative='larger')

(0.8873375320462958, 0.1874486022476365)

here, p> LOS,

null hypothesis is not rejected.

the different is not significant.

TWO TAIL TEST:

claim:

H0 mean(y) - mean(x) = 1.5

Ha mean(y) - mean(x) != 1.5 (two tail test)

In [ ]:
statsmodels.stats.weightstats.ztest(y,x,value=1.5,alternative='two-sided')

(0.8873375320462958, 0.374897204495273)

here, p> LOS,

null hypothesis is not rejected.

the difference is not significant.

In [ ]:
import os


In [ ]:
os.getcwd()

'/content'

In [ ]:
os.chdir(r'/content/drive/MyDrive/STATSCDAC')

In [ ]:
os.getcwd()

'/content/drive/MyDrive/STATSCDAC'

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_excel('MYData.xlsx')

In [ ]:
df

Sample  Value
0       A      5
1       A      4
2       A      6
3       A      7
4       A      5
5       B      6
6       B      5
7       B      4
8       B      7
9       B      6
10      C     10
11      C     14
12      C     13
13      C     12

Now the tool we use is:

**ONE WAY ANOVA**:

--> ANOVA is used when the **data is continuous**

--> ANOVA full form - ANalysis Of VAriance

--> This tool is used when we have multiple samples.

--> we can calculate mean for continuous data.


Hypotheis:

H0: all the differences of pair of means are equal

Ha: there is atleast one pair of means where the difference is not equal to zero.


ex:not for ANOVA! gen example.
 there are 4 states where the sales of cars are made. check whether the count of sales are equal?

the expected pattern is 0.25,0.25,0.25,0.25
the

In [ ]:
import statsmodels.api as sm

In [ ]:
from statsmodels.formula.api import ols

In [ ]:
import statsmodels.stats.multicomp

Creating the model:

sample --> predictor

value --> response

output --> outcome

In [ ]:
mod1 = ols('Value ~ Sample', data = df).fit()

#creating the model

In [ ]:
tb1 = sm.stats.anova_lm(mod1)

#sm.stats.anova_lm is a function that computes the ANOVA table.
#The mod1 object obtained from fitting the linear regression model is passed as the argument to the anova_lm function.
#The function computes the ANOVA table, which provides information about the sum of squares, degrees of freedom, mean squares, F-statistic, and p-values for each term in the linear regression model.
#The ANOVA table is then assigned to the variable tb1.

In [ ]:
tb1

# here the samples are A,B,C
#F stands for Fisher value, this is used in ANOVA calculations
#df - degree of freedom (sample size -1 =2)
#residual = no of rows - no of samples.

df      sum_sq    mean_sq          F    PR(>F)
Sample     2.0  130.278571  65.139286  37.416822  0.000012
Residual  11.0   19.150000   1.740909        NaN       NaN

p<0.05,

so we reject H0,

so there is atleast one pair where the difference of means is not equal to zero.

so some changes has to be made.



In [ ]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [ ]:
pairwise_tukeyhsd(df.Value,df.Sample)

In [ ]:
print(pairwise_tukeyhsd(df.Value,df.Sample))

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj   lower  upper  reject
---------------------------------------------------
     A      B      0.2 0.9689 -2.0538 2.4538  False
     A      C     6.85    0.0  4.4595 9.2405   True
     B      C     6.65    0.0  4.2595 9.0405   True
---------------------------------------------------


1st one:

mean diff between A and B is 0.2
mean diff of A and B can be -2.05 to 2.4538
0 falls between them.
H0 is not nejected becauce p is 0.9 > 0.05 and 0 falls between -2.05 and 2.4538
the difference between difference of means and zero is not statistically significant, so it can be assumed that the difference is due to chance variation. so it can be ignored, because of small value.


2nd one:

mean diff betwwn A and C is 6.85
mean diff of A and c can be 4.4595 to 9.2405
0 does not falls between them.
p < 0.05
so the differnce of means is  statistically significant, cannot be ignored, because the difference of means is not only due to chance variation but also there is some actual difference., hence cannot be ignored.

H0 is rejected, because p < 0.05

3rd one:
same as second one.

In [ ]:
print(tb1)

            df      sum_sq    mean_sq          F    PR(>F)
Sample     2.0  130.278571  65.139286  37.416822  0.000012
Residual  11.0   19.150000   1.740909        NaN       NaN


**2 WAY ANOVA:**

there will be two null hypothesis.

one is based on location,
other is based on gender.